### DISCUSSION TITLE
CadQuery exporter

### DISCUSSION THREAD

**Discussion by bat52** [jeff-dh/SolidPython#58] [Ideas]
Hi, thank you for sharing this, it seems some really nice work.
I have started looking into some relatively complex openscad project ( https://github.com/bguan/gugulele ), but after discussing with the developer, he told me that he is disappointed with openscad rendering performances, and moved to onshape FeatureScript instead.
Since I prefer the open source nature of openscad, and I am familiar with python, I like SolidPython in terms of syntax and environment.

However, the main rendered of SolidPython is still openscad, so it has the rendering issue. I implicitCAD can also be used, that claims to be better in performances, but it does not support the hull function.

After further reviewing the available projects, i bumped into CadQuery ( https://cadquery.readthedocs.io/en/latest/ ) that allows to export .step and .stl format with better performance.

So I write to ask you if you think it would be feasible to write a SolidPython extension that imports existing scad code, and exports cadquery? It seems to me the primitives should be compatible, and this should therefore also allow to generate designs in SolidPython syntax and export with the "renderer" of choice. I would say yes by browsing the code, but I told myself you probably know better.

Cheers,
Bat52

**Comment by jeff-dh** [comment]
Hey,

aaaeeehh... well....

If I understand you correctly the question whether we could plugin a different backend (as I would call it) for example cadquery, right?

I had the same idea (changing backends) a couple of times, but in the end always discarded it. The main reason is bosl2. Even though we could probably map the basics of solid python (`(cube() + sphere()) * cylinder().save_as_stl()`) with different backends (for example implicitCAD or cadquery) from my point of view `bosl2` is an existential component of SolidPython2 and we can't "transfer" bosl2 to any other backend, not even implicitCAD (that's the reason why I actually don't really care about the ImplicitCAD support at all).
That's the main reason against switching backends.

From my point of view it would probably make most sense to completely get rid of any "intermediate" code (or library) and directly use a csg library to render the models. But again, we would loose all high level (bosl2) functionality and I don't want to maintain those modeling features in SolidPython. Therefor openscad + bosl2 seems to be a "optimal" backend (a local minimum).

Furthermore I think the SolidPython "design-flow" is tied pretty close to openscad. I don't know cadquery, but I had a brief view at their examples and it is not obvious to me how to map this "SolidPython design flow" to it. Their design flow seems different to me (at least on the first glance).
The SolidPython design-flow is -- from my point of view -- directly derived from openscad. It took an evolution from openscad through SolidPython1 to what we are now talking about SolidPython2.

Btw: SolidPython can not translate openscad code to python and back to openscad. It can generate openscad code from a python "object-tree" and it can import openscad code that can be used in this object-tree, but this is more like a importing a header file. It only extracts the module signatures and create python wrappers for them to be callable from withing SolidPython. SolidPython has no idea what's actually in a module or what it actually does, it only knows the name and the parameters of imported modules. This is not enough to translate it to any other backend.

**Reply by bat52** [reply]
OK thank you very much for the quick reply and detailed explanation!

**Comment by bat52** [comment]
A quick update on this topic... I reached out the gugulele developer, and I have started contributing to a new python-based implementation that supports several backends (cadquery, blender, trimesh, solidpython2).

Here the solidpython2 code, just in case
https://github.com/bat52/pylele/blob/main/api/sp2_api.py

So basically we now have a multi-backend thingy that is capable of generating consitent basic geometries with different libraries.
Now the fun would be being able to use this basic library also with the library-specific functions (ie import a specific file type that is only supported with a specific backend).

In this context I have implemented .stl import with solidpython2: based on openscad documentation, both ascii and binary .stl import should work, but for me only binary seems to work. Have you experience that too ?

On the other hand, I would also like to import .scad files, so that we can scrap other designs from the internet... I understand from the examples solidpython supports importing .scad that include modules and functions that can be called 'natively' in python.

But... is there a method to import a scad file that simply defines a geometry (without a method) ? I know I could convert the scad to .stl, then import the .stl, but that seems a convoluted technique...

Thanks again for the support!
